crates/compiler/roar/src/ops.rs

```diff
@@ -0,0 +1,69 @@
+
+
+use crate::storage::{Constant,ByteSize,Offset,Input,Output};
+
+#[derive(Clone,Debug)]
+pub(crate) enum Sign {
+    Signed, 
+    Unsigned 
+}
+#[derive(Clone,Debug)]
+pub(crate) enum Flag {
+    Neq
+}
+type Flags = Flag;
+#[derive(Clone,Debug)]
+///All operations used, arithemetic operations take a signedness flag to increase readability
+pub(crate) enum OpCode {
+    Add(Sign),
+    Sub(Sign),
+    Mul(Sign),
+    FloatAdd,
+    FloatMul,
+    FloatDiv,
+    FloatSub,
+    Jump,
+    JumpIf(Flags),
+    ///Create a structure reference  
+    Create,
+    Load(Offset),
+    Store(Offset),
+    ///Copy `$3` number bytes from offset `$1` to offset `$2`
+    Copy(Offset,Offset,ByteSize),
+    BitAnd,
+    BitOr,
+    BitXor,
+    LogAnd,
+    LogOr,
+    LogXor,
+    ///Set the highest bit to zero (even if it is unsigned)
+    Abs,
+    ///Truncate a register, if signed, copy sign bit to new highest point
+    Trunc(Sign,ByteSize),
+    ///Extend a register, if signed, copy sign bit to new highest point, then zero out the original point
+    Extend(Sign,ByteSize),
+    ///Convert a floating point register to a regular register
+    ToInt,
+    ///Convert a regular register to a floating point register
+    ToFloat,
+    ///Copy bits from a float register to a regular register, without conversion
+    ToIntRaw,
+    ///Copy bits from a regular register to a floating point register, without conversion 
+    ToFloatRaw,
+    ///Shift left instruction, whether arithmetic or logical depends on `Sign` flag
+    ShiftLeft(Sign,ByteSize),
+    ///Shift right instruction, whether arithmetic or logical depends on `Sign` flag
+    ShiftRight(Sign,ByteSize),
+    ///Basic move instruction 
+    Move
+}
+#[derive(Clone,Debug)]
+#[non_exhaustive]
+///The actual type of operations in ROAR
+pub struct Operation {
+    pub output : Output,
+    pub opcode : OpCode,
+    pub inputs : Vec<Input>
```

Comment:
How many inputs are possible? If just two or three, I would inline them directly and use the nul register for unused inputs.

If it is only 1 or 2 ops are special (like call), it may be worth storing it's arguments sideband in a separate list. The input here would just be an index to that separate list.

If many ops are special, or just for simplicity, we can use a dynamic length container here, but we should use an arena allocator. I think the type we would want would be:
```
    pub inputs: &'a [Input],
```
And we would be using a bumpalo arena to allocate them. In our code, you will regularly see `.into_bump_slice` and `env.arena.alloc(...)`.